File: libs/workflow-core/src/graph/walker.ts
import type { WorkflowDefinition, NodeDefinition, EdgeDefinition } from '../models';
import { CelEvaluator, type CelContext } from '../cel/evaluator';
import { LoopGuard } from './loop-guard';

export type WalkerCommand =
  | { type: 'execute_action'; nodeId: string; handler: string; input: Record<string, any> }
  | { type: 'execute_workflow'; nodeId: string; workflowRef: string; workflowVersion?: number; input: Record<string, any> }
  | { type: 'fork'; nodeId: string; branches: string[]; joinNode: string }
  | { type: 'wait_for_signal'; nodeId: string; timeoutMs?: number }
  | { type: 'complete'; context: Record<string, any> }
  | { type: 'fail'; nodeId: string; reason: string };

export interface CommandResult {
  contextUpdates?: Record<string, any>;
  signalReceived?: boolean;
  error?: string;
}

export class GraphWalker {
  private ctx: Record<string, any>;
  private cursorNodeId: string;
  private nodeMap: Map<string, NodeDefinition>;
  private edgesByFrom: Map<string, EdgeDefinition[]>;
  private cel: CelEvaluator;
  private loopGuard: LoopGuard;
  private pendingError: string | null = null;
  private startNodeId: string;
  private stopNodeId: string | null = null;

  constructor(
    private definition: WorkflowDefinition,
    params: Record<string, any>,
  ) {
    this.ctx = { ...params };
    this.cursorNodeId = definition.entrypoint;
    this.startNodeId = definition.entrypoint;
    this.cel = new CelEvaluator();
    this.loopGuard = new LoopGuard(definition.maxIterations ?? 100);

    this.nodeMap = new Map();
    for (const node of definition.nodes) {
      this.nodeMap.set(node.id, node);
    }

    this.edgesByFrom = new Map();
    for (const edge of definition.edges) {
      const list = this.edgesByFrom.get(edge.from) ?? [];
      list.push(edge);
      this.edgesByFrom.set(edge.from, list);
    }
  }

  setBounds(startNodeId: string, stopNodeId: string): void {
    this.cursorNodeId = startNodeId;
    this.startNodeId = startNodeId;
    this.stopNodeId = stopNodeId;
  }

  next(): WalkerCommand {
    // Check for pending error from previous resolve
    if (this.pendingError) {
      const error = this.pendingError;
      this.pendingError = null;
      return { type: 'fail', nodeId: this.cursorNodeId, reason: error };
    }

    // Walk through decision nodes internally until we hit an actionable node
    while (true) {
      // Check for setBounds stop
      if (this.stopNodeId && this.cursorNodeId === this.stopNodeId) {
        return { type: 'complete', context: { ...this.ctx } };
      }

      const node = this.nodeMap.get(this.cursorNodeId);
      if (!node) {
        return { type: 'fail', nodeId: this.cursorNodeId, reason: `node '${this.cursorNodeId}' not found` };
      }

      switch (node.type) {
        case 'terminal':
          return { type: 'complete', context: { ...this.ctx } };

        case 'decision': {
          const nextNodeId = this.resolveDecision(node);
          if (nextNodeId === null) {
            if (this.pendingError) {
              const error = this.pendingError;
              this.pendingError = null;
              return { type: 'fail', nodeId: node.id, reason: error };
            }
            return { type: 'fail', nodeId: node.id, reason: `no matching edge from decision node '${node.id}'` };
          }
          this.cursorNodeId = nextNodeId;
          continue;
        }

        case 'action': {
          const input = this.evaluateInput(node);
          if (input === null) {
            const reason = this.pendingError ?? `CEL input evaluation failed on node '${node.id}'`;
            this.pendingError = null;
            return { type: 'fail', nodeId: node.id, reason };
          }
          return { type: 'execute_action', nodeId: node.id, handler: node.handler!, input };
        }

        case 'workflow': {
          const input = node.workflowInput ? this.evaluateInput({ ...node, input: node.workflowInput }) : {};
          if (input === null) {
            const reason = this.pendingError ?? `CEL workflowInput evaluation failed on node '${node.id}'`;
            this.pendingError = null;
            return { type: 'fail', nodeId: node.id, reason };
          }
          return {
            type: 'execute_workflow',
            nodeId: node.id,
            workflowRef: node.workflowRef!,
            workflowVersion: node.workflowVersion,
            input: input!,
          };
        }

        case 'fork':
          return {
            type: 'fork',
            nodeId: node.id,
            branches: node.branches!,
            joinNode: node.joinNode!,
          };

        case 'gate':
          return {
            type: 'wait_for_signal',
            nodeId: node.id,
            timeoutMs: node.timeoutMs,
          };

        default:
          return { type: 'fail', nodeId: node.id, reason: `unknown node type '${node.type}'` };
      }
    }
  }

  resolve(result: CommandResult): void {
    if (result.error) {
      this.pendingError = result.error;
      return;
    }

    if (result.contextUpdates) {
      Object.assign(this.ctx, result.contextUpdates);
    }

    const node = this.nodeMap.get(this.cursorNodeId)!;

    // Gate node: route based on signalReceived
    if (node.type === 'gate') {
      if (result.signalReceived) {
        this.advanceToNextEdge(node.id);
      } else if (node.timeoutEdge) {
        this.cursorNodeId = node.timeoutEdge;
      } else {
        this.pendingError = `gate '${node.id}' timed out with no timeoutEdge`;
      }
      return;
    }

    // Fork node: advance to joinNode
    if (node.type === 'fork') {
      this.cursorNodeId = node.joinNode!;
      return;
    }

    // Action, workflow: follow single outgoing edge (with loop guard)
    this.advanceToNextEdge(node.id);
  }

  getContext(): Record<string, any> {
    return { ...this.ctx };
  }

  private resolveDecision(node: NodeDefinition): string | null {
    const edges = (this.edgesByFrom.get(node.id) ?? [])
      .slice()
      .sort((a, b) => (a.priority ?? 0) - (b.priority ?? 0));

    const celContext: CelContext = { ctx: this.ctx, _loop: this.loopGuard.getLoopMap() };

    let matchedEdge: typeof edges[0] | null = null;

    for (const edge of edges) {
      if (!edge.when) {
        // Default/fallback edge
        matchedEdge = edge;
        break;
      }

      try {
        const result = this.cel.evaluateCondition(edge.when, celContext);
        if (result) {
          matchedEdge = edge;
          break;
        }
      } catch (e: any) {
        this.pendingError = `CEL expression on edge '${edge.from}->${edge.to}' failed: ${e.message}`;
        return null;
      }
    }

    if (!matchedEdge) return null;

    // Only check loop guard on the edge actually taken
    const allowed = this.loopGuard.check(matchedEdge.from, matchedEdge.to, matchedEdge.maxIterations);
    if (!allowed) {
      this.pendingError = `loop limit exceeded on edge '${matchedEdge.from}->${matchedEdge.to}'`;
      return null;
    }

    return matchedEdge.to;
  }

  private evaluateInput(node: NodeDefinition): Record<string, any> | null {
    if (!node.input) return {};

    const celContext: CelContext = { ctx: this.ctx, _loop: this.loopGuard.getLoopMap() };
    try {
      return this.cel.evaluateInput(node.input, celContext);
    } catch (e: any) {
      this.pendingError = `CEL input on node '${node.id}' failed: ${e.message}`;
      return null;
    }
  }

  private advanceToNextEdge(nodeId: string): void {
    const edges = this.edgesByFrom.get(nodeId) ?? [];
    const edge = edges[0];
    if (!edge) return;

    // Check loop guard on non-decision edge traversals too
    const allowed = this.loopGuard.check(edge.from, edge.to, edge.maxIterations);
    if (!allowed) {
      this.pendingError = `loop limit exceeded on edge '${edge.from}->${edge.to}'`;
      return;
    }

    this.cursorNodeId = edge.to;
  }
}

File: apps/controlplane/internal/graph/walker.go
package graph

import (
	"fmt"
	"sort"

	"github.com/adamgilman/fishplate/apps/controlplane/internal/cel"
	"github.com/adamgilman/fishplate/apps/controlplane/internal/model"
)

type CommandType string

const (
	CmdExecuteAction   CommandType = "execute_action"
	CmdExecuteWorkflow CommandType = "execute_workflow"
	CmdFork            CommandType = "fork"
	CmdWaitForSignal   CommandType = "wait_for_signal"
	CmdComplete        CommandType = "complete"
	CmdFail            CommandType = "fail"
)

type WalkerCommand struct {
	Type            CommandType
	NodeID          string
	Handler         string
	Input           map[string]any
	WorkflowRef     string
	WorkflowVersion int
	Branches        []string
	JoinNode        string
	TimeoutMs       int
	Context         map[string]any // for CmdComplete
	Reason          string         // for CmdFail
}

type CommandResult struct {
	ContextUpdates map[string]any
	SignalReceived bool
	Error          string
}

type Walker struct {
	def        *model.WorkflowDefinition
	ctx        map[string]any
	cursor     string
	nodeMap    map[string]*model.NodeDefinition
	edgesByFrom map[string][]model.EdgeDefinition
	loopGuard  *LoopGuard
	cel        *cel.Evaluator
	pendingErr string
	lastNode   string // last node yielded by Next()
}

func NewWalker(def *model.WorkflowDefinition, params map[string]any) *Walker {
	ctx := make(map[string]any)
	for k, v := range params {
		ctx[k] = v
	}

	nodeMap := make(map[string]*model.NodeDefinition, len(def.Nodes))
	for i := range def.Nodes {
		nodeMap[def.Nodes[i].ID] = &def.Nodes[i]
	}

	edgesByFrom := make(map[string][]model.EdgeDefinition)
	for _, e := range def.Edges {
		edgesByFrom[e.From] = append(edgesByFrom[e.From], e)
	}

	maxIter := def.MaxIterations
	if maxIter <= 0 {
		maxIter = 100
	}

	return &Walker{
		def:         def,
		ctx:         ctx,
		cursor:      def.Entrypoint,
		nodeMap:     nodeMap,
		edgesByFrom: edgesByFrom,
		loopGuard:   NewLoopGuard(maxIter),
		cel:         cel.NewEvaluator(),
	}
}

func (w *Walker) Next() WalkerCommand {
	if w.pendingErr != "" {
		reason := w.pendingErr
		w.pendingErr = ""
		return WalkerCommand{Type: CmdFail, Reason: reason}
	}

	for i := 0; i < w.def.MaxIterations; i++ {
		node, ok := w.nodeMap[w.cursor]
		if !ok {
			return WalkerCommand{Type: CmdFail, Reason: fmt.Sprintf("node %q not found", w.cursor)}
		}

		edges := w.edgesByFrom[w.cursor]

		switch node.Type {
		case model.NodeTypeTerminal:
			if len(edges) == 0 {
				return WalkerCommand{Type: CmdComplete, Context: w.copyCtx()}
			}
			// Terminal with edges = entrypoint, follow edge
			next, err := w.followEdge(edges)
			if err != nil {
				return WalkerCommand{Type: CmdFail, Reason: err.Error()}
			}
			w.cursor = next
			continue

		case model.NodeTypeDecision:
			next, err := w.followEdge(edges)
			if err != nil {
				return WalkerCommand{Type: CmdFail, Reason: err.Error()}
			}
			w.cursor = next
			continue

		case model.NodeTypeAction:
			w.lastNode = node.ID
			cmd := WalkerCommand{
				Type:    CmdExecuteAction,
				NodeID:  node.ID,
				Handler: node.Handler,
			}
			if node.Input != "" {
				input, err := w.cel.EvaluateInput(node.Input, w.ctx, w.loopGuard.GetLoopMap())
				if err != nil {
					return WalkerCommand{Type: CmdFail, Reason: fmt.Sprintf("CEL input eval: %v", err)}
				}
				cmd.Input = input
			}
			if node.TimeoutMs > 0 {
				cmd.TimeoutMs = node.TimeoutMs
			}
			return cmd

		case model.NodeTypeWorkflow:
			w.lastNode = node.ID
			cmd := WalkerCommand{
				Type:            CmdExecuteWorkflow,
				NodeID:          node.ID,
				WorkflowRef:     node.WorkflowRef,
				WorkflowVersion: node.WorkflowVersion,
			}
			if node.WorkflowInput != "" {
				input, err := w.cel.EvaluateInput(node.WorkflowInput, w.ctx, w.loopGuard.GetLoopMap())
				if err != nil {
					return WalkerCommand{Type: CmdFail, Reason: fmt.Sprintf("CEL input eval: %v", err)}
				}
				cmd.Input = input
			}
			return cmd

		case model.NodeTypeFork:
			w.lastNode = node.ID
			return WalkerCommand{
				Type:     CmdFork,
				NodeID:   node.ID,
				Branches: node.Branches,
				JoinNode: node.JoinNode,
			}

		case model.NodeTypeGate:
			w.lastNode = node.ID
			cmd := WalkerCommand{
				Type:   CmdWaitForSignal,
				NodeID: node.ID,
			}
			if node.TimeoutMs > 0 {
				cmd.TimeoutMs = node.TimeoutMs
			}
			return cmd

		default:
			return WalkerCommand{Type: CmdFail, Reason: fmt.Sprintf("unknown node type %q", node.Type)}
		}
	}

	return WalkerCommand{Type: CmdFail, Reason: "max iterations exceeded"}
}

func (w *Walker) Resolve(result CommandResult) {
	if result.Error != "" {
		w.pendingErr = result.Error
		return
	}

	// Merge context updates
	for k, v := range result.ContextUpdates {
		w.ctx[k] = v
	}

	node, ok := w.nodeMap[w.lastNode]
	if !ok {
		w.pendingErr = fmt.Sprintf("resolve: node %q not found", w.lastNode)
		return
	}

	switch node.Type {
	case model.NodeTypeGate:
		if !result.SignalReceived {
			// Stay at gate, don't advance
			return
		}
		w.advanceCursor()
	case model.NodeTypeFork:
		if node.JoinNode != "" {
			w.cursor = node.JoinNode
		} else {
			w.advanceCursor()
		}
	default:
		w.advanceCursor()
	}
}

func (w *Walker) advanceCursor() {
	edges := w.edgesByFrom[w.lastNode]
	if len(edges) > 0 {
		w.cursor = edges[0].To
	}
}

func (w *Walker) followEdge(edges []model.EdgeDefinition) (string, error) {
	// Sort by priority
	sorted := make([]model.EdgeDefinition, len(edges))
	copy(sorted, edges)
	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i].Priority < sorted[j].Priority
	})

	loopMap := w.loopGuard.GetLoopMap()

	for _, edge := range sorted {
		if edge.When != "" {
			result, err := w.cel.EvaluateCondition(edge.When, w.ctx, loopMap)
			if err != nil {
				return "", fmt.Errorf("CEL condition eval on edge %s->%s: %v", edge.From, edge.To, err)
			}
			if !result {
				continue
			}
		}
		// Match found, check loop guard
		if !w.loopGuard.Check(edge.From, edge.To, edge.MaxIterations) {
			return "", fmt.Errorf("loop limit exceeded on edge %s->%s", edge.From, edge.To)
		}
		return edge.To, nil
	}

	return "", fmt.Errorf("no matching edge from %q", edges[0].From)
}

func (w *Walker) copyCtx() map[string]any {
	m := make(map[string]any, len(w.ctx))
	for k, v := range w.ctx {
		m[k] = v
	}
	return m
}
